packages/openactive-integration-tests/test/report-generator.js

```diff
@@ -392,6 +424,83 @@ class LoggerGroup {
     else return "passed";
   }
 
+  /**
+   * Get stats about which tests have OpportunityNotFound events and thus
+   * failed due to missing required opportunity data.
+   */
+  get missingOpportunityDataStats() {
+    if (this._missingOpportunityDataStats) { return this._missingOpportunityDataStats; }
+    const allAugmentedEvents = this.loggers.flatMap((logger) => {
+      const testConfig = _.pick(logger, [
+        'opportunityTypeName',
+
+        'featureName',
+        'implementedDisplayLabel',
+        'suiteName',
+        'htmlLocalPath',
+      ]);
+      const flowStageLogs = Object.values(logger.flow);
+      return flowStageLogs.flatMap(flowStageLog => (
+        flowStageLog.events
+          .filter(event => event.type === 'OpportunityNotFound')
+          // Associate each of these events with a specific test
+          .map(event => ({ ..._.omit(event, ['type']), testConfig }))
+      ));
+    });
+    const uniqueAugmentedEvents = _.uniqWith(allAugmentedEvents, _.isEqual);
+    this._missingOpportunityDataStats = uniqueAugmentedEvents;
+    return this._missingOpportunityDataStats;
+  }
+
+  /**
+   * Groups the results from `missingOpportunityDataStats` into something
+   * that can be shown in the summary report.
```

Comment:
@nickevansuk this is similar to my initial implementation for this, but the issue was that the output would very quickly become huge. Even just for the `payment` category of tests (let alone the full test run!), the missing data section was several pages long and which is just number counts for each combination, which one can easily get lost in.

I had a chat with Civ about this as well and we agreed that it would be good to get this information across, but difficult at present. And, regardless, it's currently possible to get this information by running Test Data Generator (which is now recommended in the output text if there is any missing data).

Besides that, the main use case which we're aware of from our experience is when a data publisher has simply forgotten to run some script or do some action which generates loads of random test data. So in those sorts of cases, they just need a rough idea that they need to generate more data.

In future, perhaps we could have something like the summarised view as it is but which links to a more complicated file, which perhaps is just generated using test-data-generator (DRY) to shortcut the journey for the user.

Reply:
Great thinking! Responding to this late via a new issue here: https://github.com/openactive/openactive-test-suite/issues/684